main.go

```diff
@@ -18,5 +18,223 @@
+	"github.com/networkservicemesh/sdk/pkg/tools/token"
+	"github.com/networkservicemesh/sdk/pkg/tools/tracing"
+)
+
+const registryAddr = "registry:5002"
+
+func handleError(w http.ResponseWriter, err error) {
+	fmt.Println("Error: ", err)
+	http.Error(w, "Error encoding JSON", http.StatusInternalServerError)
+}
+
 func main() {
+	ctx, cancel := signal.NotifyContext(
+		context.Background(),
+		os.Interrupt,
+		syscall.SIGHUP,
+		syscall.SIGTERM,
+		syscall.SIGQUIT,
+	)
+	defer cancel()
+
+	// Setup logger
+	log.EnableTracing(true)
+	logrus.SetFormatter(&nested.Formatter{})
+	ctx = log.WithLog(ctx, logruslogger.New(ctx, map[string]interface{}{"cmd": os.Args[:1]}))
+	logger := log.FromContext(ctx)
+
+	// Setup Spire
+	source, err2 := workloadapi.NewX509Source(ctx)
+	if err2 != nil {
+		logger.Fatalf("Error getting x509 source: %v", err2.Error())
+	}
+	var svid *x509svid.SVID
+	svid, err2 = source.GetX509SVID()
+	if err2 != nil {
+		logger.Fatalf("Error getting x509 svid: %v", err2.Error())
+	}
+	logger.Infof("sVID: %q", svid.ID)
+
+	tlsClientConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
+	tlsClientConfig.MinVersion = tls.VersionTLS12
+
+	// Configure and run REST API server
+	configureAndRunRESTServer(logger)
+
+	// Create gRPC dial options
+	dialOptions := append(tracing.WithTracingDial(),
+		grpc.WithDefaultCallOptions(
+			grpc.WaitForReady(true),
+			grpc.PerRPCCredentials(token.NewPerRPCCredentials(spiffejwt.TokenGeneratorFunc(source, time.Duration(10)*time.Minute))),
+		),
+		grpc.WithTransportCredentials(credentials.NewTLS(tlsClientConfig)),
+	)
+
+	// Monitor Connections
+	for ; ctx.Err() == nil; time.Sleep(time.Millisecond * 100) {
+		var nseChannel = getNseChannel(ctx, logger, dialOptions)
+
+		for ctx.Err() == nil {
+			select {
+			case <-ctx.Done():
+				return
+			case nse, ok := <-nseChannel:
+				if !ok {
+					break
+				}
+
+				nsmgrAddr := strings.Join(strings.Split(strings.TrimPrefix(nse.GetNetworkServiceEndpoint().Url, "tcp://["), "]:"), ":")
+
+				if !nsmgrs[nsmgrAddr] {
+					nsmgrs[nsmgrAddr] = true
+					logger.Infof("Extracted NSMGR addr: %q", nsmgrAddr)
+
+					// Start a goroutine for each nsmgr
+					go func() {
+						nsmgrConn, err6 := grpc.Dial(nsmgrAddr, dialOptions...)
+						if err6 != nil {
+							logger.Errorf("Failed dial to NSMgr %q: %v", nsmgrAddr, err6.Error())
+							return
+						}
+
+						clientConnections := networkservice.NewMonitorConnectionClient(nsmgrConn)
+
+						stream, err7 := clientConnections.MonitorConnections(ctx, &networkservice.MonitorScopeSelector{})
+						if err7 != nil {
+							logger.Errorf("Error from MonitorConnectionClient: %v", err7.Error())
+							return
+						}
+
+						for ctx.Err() == nil {
+							event, err8 := stream.Recv()
+							if err8 != nil {
+								logger.Errorf("Error from %q nsmgr monitorConnection stream: %v", nsmgrAddr, err8.Error())
+								time.Sleep(time.Millisecond * 500)
```

Comment:
Could you please remind why we need a delay here?

Reply:
In this section, the nsmgr goroutine work (which probably restarted with a different address) is completed and all connections that were monitored on it are deleted from the storage. At the same time, the connections change their ID, as their path segment changes. These connections can be duplicated on other managers (if they pass through 2 forwarders), from which UPDATE events come for these connections. If we do not delay, the connections will first be deleted, then the UPDATE events will restore them, and data duplication will occur in the storage. So the delay before cleaning up is needed to prevent data duplication.